android/src/main/java/io/grpc/android/AndroidChannelBuilder.java

```diff
@@ -0,0 +1,255 @@
+  }
+
+  /** Creates a new builder for the given host, port, and Android context. */
+  public static AndroidChannelBuilder forAddress(String name, int port, Context context) {
+    return forTarget(GrpcUtil.authorityFromHostAndPort(name, port), context);
+  }
+
+  private AndroidChannelBuilder(String target, Context context) {
+    delegateBuilder = OkHttpChannelBuilder.forTarget(target);
+    this.context = context;
+  }
+
+  @Override
+  protected ManagedChannelBuilder<?> delegate() {
+    return delegateBuilder;
+  }
+
+  @Override
+  public ManagedChannel build() {
+    return new AndroidChannel(delegateBuilder.build(), context);
+  }
+
+  /**
+   * Wraps an OkHttp channel and handles invoking the appropriate methods (e.g., {@link
+   * ManagedChannel#resetConnectBackoff}) when the device network state changes.
+   */
+  @VisibleForTesting
+  static final class AndroidChannel extends ManagedChannel {
+
+    private final ManagedChannel delegate;
+    private final Context context;
+    private final ConnectivityManager connectivityManager;
+
+    private DefaultNetworkCallback defaultNetworkCallback;
+    private NetworkReceiver networkReceiver;
+
+    private final Object lock = new Object();
+
+    // May only go from true to false, and lock must be held when assigning this
+    private volatile boolean needToUnregisterListener = true;
+
+    @VisibleForTesting
+    AndroidChannel(final ManagedChannel delegate, Context context) {
+      this.delegate = delegate;
+      this.context = context;
+      connectivityManager =
+          (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
+
+      // Android N added the registerDefaultNetworkCallback API to listen to changes in the device's
+      // default network. For earlier Android API levels, use the BroadcastReceiver API.
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && connectivityManager != null) {
+        NetworkInfo currentNetwork = connectivityManager.getActiveNetworkInfo();
+
+        // The connection status may change before registration of the listener is complete, but
+        // this will at worst result in invoking resetConnectBackoff() instead of enterIdle() (or
+        // vice versa) on the first network change.
+        boolean isConnected = currentNetwork != null && currentNetwork.isConnected();
+
+        defaultNetworkCallback = new DefaultNetworkCallback(isConnected);
+        connectivityManager.registerDefaultNetworkCallback(defaultNetworkCallback);
+      } else {
+        networkReceiver = new NetworkReceiver();
+        IntentFilter networkIntentFilter =
+            new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION);
+        context.registerReceiver(networkReceiver, networkIntentFilter);
+      }
+    }
+
+    private void unregisterNetworkListener() {
+      if (needToUnregisterListener) {
+        synchronized (lock) {
+          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && connectivityManager != null) {
```

Comment:
Does repeating the `Build` check here help dead code elimination? If not, is it necessary?

Reply:
Refactored to use runnable instead of this conditional